src/Context/ContextHandlerTrait.php

```diff
@@ -46,6 +46,14 @@ protected function mapContext(CoreContextAwarePluginInterface $plugin, RulesStat
       // Check if a data selector is configured that maps to the state.
       if (isset($this->configuration['context_mapping'][$name])) {
         $typed_data = $state->applyDataSelector($this->configuration['context_mapping'][$name]);
+
+        if ($typed_data->getValue() === NULL && !$definition->isAllowedNull()) {
```

Comment:
It's not sufficient to check this for mapping only, we need to check it regardless where the data comes from.